### PR TITLE
Expose selectionHeightStyle and selectionWidthStyle on TextFields

### LINKED ISSUE #39755: Selection of any justified-text is inaccurate in non-latin languages
I hope I can clarify this issue because it is a little weird. It took me a while to figure out when exactly does it occur.

Here is a very primitive application:

```dart
// main.dart
import 'package:flutter/material.dart';
void main() => runApp(MaterialApp(
  home: Scaffold(
    appBar: AppBar(),
    body: Padding(
      padding: const EdgeInsets.all(18.0),
      child: TextField(
        textAlign: TextAlign.justify,
        maxLines: null,
      ),
    ),
  ),
));
```

Now, type any latin charachter text. Try to select any single or multiple words, you will find that the selection blue highlight matches accurately the boundaries of the word(s). See this screen shot:

image

_____________________

Then, put some Korean text. You will find that some words are poorly highlighted:

image

> 한국어(韓國語)는 대한민국과 조선민주주의인민공화국의 공용어로, 대한민국에서는 한국어 또는 한국말이라고 부르고, 조선민주주의인민공화국에서는 조선말이라고 부른다.

_____________________

Also in Arabic texts (it does not matter here that the `textDirection` is not RTL. Both RTL and LTR have the same issue):

image

> العربية لغة رسمية في كل دول الوطن العربي إضافة إلى كونها لغة رسمية في تشاد وإريتريا وإسرائيل. وهي إحدى اللغات الرسمية الست في منظمة الأمم المتحدة، ويُحتفل باليوم العالمي للغة العربية في 18 ديسمبر كذكرى اعتماد العربية بين لغات العمل في الأمم المتحدة.

_____________________

Same in Hebrew:

image

> עִבְרִית היא שפה שמית, ממשפחת השפות האפרו-אסיאתיות, הידועה כשפתם של היהודים ושל השומרונים, אשר ניב מודרני שלה (עברית ישראלית) הוא שפתה הרשמית של מדינת ישראל, מעמד שעוגן בשנת 2018 בחוק יסוד: ישראל – מדינת הלאום של העם היהודי.

_____________________

Just copy these texts and paste them in the input field, then try to select one word. In all cases above I just long tapped on one word and the highlight missed the word boundry as shown.

It does have to be a text field or input though. Any justified selectable text would reproduce the same behavior. I tried in both stable and master channels.

Here is my `flutter doctor` output:

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, v1.9.8-pre.54, on Linux, locale en_US.UTF-8)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Android Studio (version 3.5)
[✓] VS Code (version 1.37.1)
[✓] Connected device (1 available)

• No issues found!
```

### PR DESCRIPTION
**Pull request by GaryQian**
## Description

This adds `selectionHeightStyle` and `selectionWidthStyle` to TextField and CupertinoTextField. These properties are passed to GetBoxesForRange as `boxHeightStyle` and `boxWidthStyle`.

This allows for text selection/highlighting to use different kinds of boxes for the selection.

This has been possible for a long time now in getBoxesForRange, but the most obvious application is only now being exposed via this PR.

## Related Issues

Adds features for https://github.com/flutter/flutter/issues/31459

Fixes https://github.com/flutter/flutter/issues/39755

## Tests

Goldens added to verify correct highlights are used.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

**Review by justinmc — APPROVED** [review]
LGTM

Can you explain the end goal of adding these parameters for my own curiosity?  I see what they do, but I'm wondering if Flutter users want to control this, or if this is being added to fix some specific bug or enable some other feature.

**Comment by GaryQian** [comment]
The directly linked issue doesn't do a great job at explaining the actual use cases that people were asking for. For example, in https://github.com/flutter/flutter/issues/21679#issuecomment-485456967 reveals an actually quite common use case that without this API is actually impossible to get looking correct. Whenever multiple languages and fonts mix (and sometimes even cases within a single languages), the highlighting can become very misaligned, making the selection look terrible. This also is important in replicating behavior in iOS vs android vs Windows etc. These platforms do different things with the end-of-line behavior, some choosing wide while others choose tight.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 1]
```diff
@@ -196,10 +198,15 @@ class CupertinoTextField extends StatefulWidget {
   ///
   /// If specified, the [maxLength] property must be greater than zero.
   ///
+  /// The [selectionHeightStyle] and [selectionWidthStyle] properties allow
+  /// changing the shape of the selection highlighting. These properties default
+  /// to [ui.BoxHeightStyle.tight] and [ui.BoxWidthStyle.tight] respectively and
+  /// must not be null.
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/text_field.dart]: add asserts in the constructor below that these are not null?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 2]
```diff
@@ -530,6 +539,16 @@ class CupertinoTextField extends StatefulWidget {
   /// and [CupertinoColors.activeOrange] in the dark theme.
   final Color cursorColor;
 
+  /// Controls how tall the selection highlight boxes are computed to be.
+  ///
+  /// See [ui.BoxHeightStyle] for details on available styles.
+  final ui.BoxHeightStyle selectionHeightStyle;
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/text_field.dart]: (unrelated to this PR, so please ignore) I wish the enum doc had actual images showing off the difference.

File: packages/flutter/lib/src/material/text_field.dart [hunk 3]
```diff
@@ -279,9 +281,15 @@ class TextField extends StatefulWidget {
   /// The text cursor is not shown if [showCursor] is false or if [showCursor]
   /// is null (the default) and [readOnly] is true.
   ///
+  /// The [selectionHeightStyle] and [selectionWidthStyle] properties allow
+  /// changing the shape of the selection highlighting. These properties default
+  /// to [ui.BoxHeightStyle.tight] and [ui.BoxWidthStyle.tight] respectively and
+  /// must not be null.
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/text_field.dart]: assert on this.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 4]
```diff
@@ -389,6 +389,8 @@ class EditableText extends StatefulWidget {
     this.cursorOpacityAnimates = false,
     this.cursorOffset,
     this.paintCursorAboveText = false,
+    this.selectionHeightStyle = ui.BoxHeightStyle.tight,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/editable_text.dart]: Is null ok? If not, assert and also update the doc comment above?

**Comment by justinmc** [comment]
Ah that comment makes it clear, thank you for explaining!  Makes sense to me.

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by goderbauer** [comment]
Looks like cirrus is unhappy about some golden file changes...

**Pull request merged**